File: mobile/android/protocol/src/main/kotlin/com/example/kinglouie/protocol/AppCore.kt
package com.example.kinglouie.protocol

import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.Signature
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec
import java.time.Instant
import java.util.UUID

enum class AppMode { WELCOME, DEMO, LIVE }

/** The only way the app obtains a network client. Demo mode never gets one: the factory refuses, and `built` proves it in tests. */
class RelayClientFactory<C>(private val make: () -> C) {
    var built = 0
        private set

    fun client(mode: AppMode): C? {
        if (mode != AppMode.LIVE) return null
        built += 1
        return make()
    }
}

/** Three pretend nodes with in-app keys and a software phone key, shown under a "Demo" banner. Nothing here touches the network. */
class DemoFleet(names: List<String> = listOf("gpu-box", "laptop", "web-01")) {
    private class Node(val pin: NodePin, val keys: KeyPair)

    private val nodes: List<Node> = names.map { name ->
        val keys = KeyPairGenerator.getInstance("Ed25519").generateKeyPair()
        val spki = keys.public.encoded
        val raw = spki.copyOfRange(spki.size - 32, spki.size)
        Node(NodePin(Identifiers.nodeId(raw), name, Hex.encode(spki)), keys)
    }

    private val deviceKeys: KeyPair = KeyPairGenerator.getInstance("EC").run {
        initialize(ECGenParameterSpec("secp256r1"))
        generateKeyPair()
    }

    val pins: List<NodePin> get() = nodes.map { it.pin }

    val deviceId: String
        get() {
            val pub = deviceKeys.public as ECPublicKey
            return Identifiers.deviceId(byteArrayOf(0x04) + P256.coordinate(pub.w.affineX) + P256.coordinate(pub.w.affineY))
        }

    /** A node-signed request, as a real node would send it. */
    fun request(index: Int, command: String, now: Instant = Instant.now()): Envelope {
        val node = nodes[index]
        val action = JsonObject(
            mapOf(
                "kind" to jsonString("tool"),
                "name" to jsonString("Bash"),
                "params" to JsonObject(mapOf("command" to jsonString(command))),
                "cwd" to jsonString("/srv/site"),
                "summary" to jsonString(summary("Bash($command)"))
            )
        )
        val message = JsonObject(
            mapOf(
                "v" to jsonNumber("1"),
                "type" to jsonString("kl.approval.request"),
                "request_id" to jsonString(UUID.randomUUID().toString()),
                "node_id" to jsonString(node.pin.id),
                "node_name" to jsonString(node.pin.name),
                "action" to action,
                "action_hash" to jsonString(Digest.sha256B64url(Jcs.bytes(action))),
                "origin" to JsonObject(mapOf("client" to jsonString("demo"), "session" to JsonNull, "job_id" to JsonNull)),
                "created_at" to jsonString(Timestamps.string(now)),
                "expires_at" to jsonString(Timestamps.string(now.plusSeconds(300))),
                "nonce" to jsonString(Messages.randomNonce())
            )
        )
        return Envelope.seal(message, node.pin.id, "Ed25519") { bytes ->
            Signature.getInstance("Ed25519").run { initSign(node.keys.private); update(bytes); sign() }
        }
    }

    /** The demo software key signs demo answers (raw r||s); it is dropped with the fleet when the owner leaves demo. */
    fun sign(bytes: ByteArray): ByteArray = Signature.getInstance("SHA256withECDSA").run {
        initSign(deviceKeys.private)
        update(bytes)
        P1363.fromDer(sign())
    }

    companion object {
        /** A summary is at most 300 code points, cut with "…" (the node's cutSummary). */
        fun summary(text: String): String {
            val cps = text.codePoints().toArray()
            return if (cps.size > 300) String(cps, 0, 299) + "…" else text
        }
    }
}

File: mobile/ios/KLProtocol/Sources/KLProtocol/AppCore.swift
import Foundation
#if canImport(CryptoKit)
import CryptoKit
#else
import Crypto
#endif

public enum AppMode: String, Codable, Equatable {
    case welcome
    case demo
    case live
}

/// The only way the app obtains a network client. Demo mode never gets one:
/// the factory refuses, and `built` proves it in tests.
public final class RelayClientFactory<Client> {
    public private(set) var built = 0
    private let make: () -> Client

    public init(make: @escaping () -> Client) {
        self.make = make
    }

    public func client(for mode: AppMode) -> Client? {
        guard mode == .live else { return nil }
        built += 1
        return make()
    }
}

/// Three pretend nodes with in-app keys and a software phone key, shown
/// under a "Demo" banner. Nothing here touches the network.
public final class DemoFleet {
    public struct Node {
        public let pin: NodePin
        fileprivate let key: Curve25519.Signing.PrivateKey
    }

    public let nodes: [Node]
    private let deviceKey = P256.Signing.PrivateKey()

    public init(names: [String] = ["gpu-box", "laptop", "web-01"]) {
        nodes = names.map { name in
            let key = Curve25519.Signing.PrivateKey()
            let raw = key.publicKey.rawRepresentation
            return Node(pin: NodePin(id: Identifiers.nodeId(ed25519Raw: raw), name: name, key: Identifiers.ed25519SpkiPrefix + Hex.encode(raw)), key: key)
        }
    }

    public var pins: [NodePin] { nodes.map { $0.pin } }

    public var deviceId: String {
        Identifiers.deviceId(raw: deviceKey.publicKey.x963Representation)
    }

    /// A node-signed request, as a real node would send it.
    public func request(on index: Int, command: String, now: Date = Date()) throws -> Envelope {
        let node = nodes[index]
        let action: JSONValue = .object([
            "kind": .string("tool"),
            "name": .string("Bash"),
            "params": .object(["command": .string(command)]),
            "cwd": .string("/srv/site"),
            "summary": .string(Self.summary("Bash(\(command))"))
        ])
        let message: JSONValue = .object([
            "v": .number("1"),
            "type": .string("kl.approval.request"),
            "request_id": .string(UUID().uuidString.lowercased()),
            "node_id": .string(node.pin.id),
            "node_name": .string(node.pin.name),
            "action": action,
            "action_hash": .string(Digest.sha256B64url(JCS.data(action))),
            "origin": .object(["client": .string("demo"), "session": .null, "job_id": .null]),
            "created_at": .string(Timestamps.string(now)),
            "expires_at": .string(Timestamps.string(now.addingTimeInterval(300))),
            "nonce": .string(Messages.randomNonce())
        ])
        return try Envelope.seal(message, alg: "Ed25519", kid: node.pin.id) { bytes in
            try node.key.signature(for: bytes)
        }
    }

    /// The software key signs demo answers; it is deleted when the owner leaves demo.
    public func sign(_ data: Data) throws -> Data {
        try deviceKey.signature(for: data).rawRepresentation
    }

    /// A summary is at most 300 code points, cut with "…" (the node's cutSummary).
    static func summary(_ text: String) -> String {
        let scalars = Array(text.unicodeScalars)
        guard scalars.count > 300 else { return text }
        var cut = String.UnicodeScalarView()
        cut.append(contentsOf: scalars[0..<299])
        cut.append("\u{2026}")
        return String(cut)
    }
}
